### PR TITLE
SCP filename validation docs

### LINKED ISSUE #52640: scp: fail on newer OpenSSH builds -  protocol error: filename does not match request
##### SUMMARY
A newer release of OpenSSH has included the commit https://github.com/openssh/openssh-portable/commit/3d896c157c722bc47adca51a58dca859225b5874 which adds validation around the file name returned by the server. This causes issues when scp is used to fetch a file as Ansible will use single quotes around the path causing a validation failure.

We've currently started to see this issue in https://github.com/ansible/ansible/pull/47732 and have added a workaround by setting `ansible_scp_extra_args: -T` to get the test working again but this should be removed once this has been fixed.

This does not necessarily have to be handled in Ansible, this issue is just to keep track of the issue until further details can be gotten from OpenSSH. If left unfixed we may start seeing these issues the next time we build a new test container, currently the `default` container is affected.

Potentially related to https://bugzilla.mindrot.org/show_bug.cgi?id=2966.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
ssh

##### ANSIBLE VERSION
```paste below
devel
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
Linux with the latest OpenSSH build. This is something that affects the scp client and not the scp version on the target.

##### STEPS TO REPRODUCE
You can easily test this out by running a container that installs the latest nightly of OpenSSH.

```
# Create file with a space in it
touch "/tmp/file space.txt"

# Start an interactive container
docker run -it ubuntu:16.04 /bin/bash

# Install Ansible and compile the latest OpenSSH version
apt-get update -y
apt-get install software-properties-common
apt-add-repository --yes --update ppa:ansible/ansible
apt-get install -y ansible build-essential zlib1g-dev libssl-dev wget sshpass
mkdir /var/lib/sshd
chmod -R 700 /var/lib/sshd/
chown -R root:sys /var/lib/sshd/
useradd -r -U -d /var/lib/sshd/ -c "sshd privsep" -s /bin/false sshd
wget -c http://www.mindrot.org/openssh_snap/openssh-SNAP-20190220.tar.gz
tar -xzf openssh-SNAP-20190220.tar.gz
cd openssh
./configure --with-privsep-path=/var/lib/sshd/ --sysconfdir=/etc/ssh 
make
make install

echo "host  ansible_host=host.docker.internal ansible_user=<username>" > inventory.ini
ANSIBLE_HOST_KEY_CHECKING=False ANSIBLE_SCP_IF_SSH=True ansible -i inventory.ini host -k -m fetch -a "src='/tmp/file space.txt' dest=/tmp/file.txt flat=yes"
```

##### EXPECTED RESULTS
The file is fetched without any issues

##### ACTUAL RESULTS
```paste below
root@d8d1f97503dc:/openssh# ANSIBLE_HOST_KEY_CHECKING=False ANSIBLE_SCP_IF_SSH=True ansible -i inventory.ini host -k -m fetch -a "src='/tmp/file space.txt' dest=/tmp/file.txt flat=yes"
SSH password: 
host | FAILED! => {
    "msg": "failed to transfer file to /tmp/file space.txt /tmp/file.txt:\n\n/etc/ssh/ssh_config line 55: Unsupported option \"gssapiauthentication\"\r\n/etc/ssh/ssh_config line 56: Unsupported option \"gssapidelegatecredentials\"\r\nprotocol error: filename does not match request\n"
}
```

You can manually run the scp command outside of Ansible to verify that this is what scp is returning. More specifically it's the fact that the path of the remote file is enclosed in single quotes like `'[host.docker.internal]:'"'"'/tmp/file space.txt'"'"''`

### PR DESCRIPTION
**Pull request by jborean93**
##### SUMMARY
Add official docs for https://github.com/ansible/ansible/issues/52640. If we start to get multiple issues then we will need to handle this error in the code and link to the entry on a failure but we are just starting with having docs for now.

Fixes https://github.com/ansible/ansible/issues/52640.

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
ssh
scp

### CODE REVIEW

File: docs/docsite/rst/reference_appendices/faq.rst [hunk 1]
```diff
@@ -643,6 +643,34 @@ but you can still access the original via ``hostvars``::
...
+
+    failed to transfer file to /tmp/ansible/file.txt\r\nprotocol error: filename does not match request
+
+This is due to a bug in newer releases of OpenSSH in the ``scp`` client that is used on the Ansible controller. These
+releases attempt to validate that the path of the file to fetch matches the requested requested path. This validation
+will fail if the remote filename requires quotes to escape spaces or non-ascii chars in its path. If you come across
+this error there are a few different ways to ignore this validation, such as:
+
+* Don't use the ``scp`` transfer mechanism by
+    * Not setting ``scp_if_ssh``, by default this uses the ``smart`` mechanism which tried SFTP first,
```
Comment (mattclay) [inline docs/docsite/rst/reference_appendices/faq.rst]:
```suggestion
    * Not setting ``scp_if_ssh``, by default this uses the ``smart`` mechanism which tries SFTP first,
```

File: docs/docsite/rst/reference_appendices/faq.rst [hunk 2]
```diff
@@ -643,6 +643,34 @@ but you can still access the original via ``hostvars``::
...
+    * Explicitly setting the host var ``ansible_scp_if_ssh=smart`` or ``ansible_scp_if_ssh=False``,
+    * Setting the environment variable ``ANSIBLE_SCP_IF_SSH`` when running Ansible to either ``smart`` or ``False``, or
+    * Setting ``scp_if_ssh`` to either ``smart`` or ``False`` under the ``[ssh_connection]`` config section in ``ansible.cfg``
+* If SCP is needed then set the ``-T`` arg to tell the SCP client to ignore path validation by
+    * Setting the var ``ansible_scp_extra_args=-T``,
+    * Setting the environment variable ``ANSIBLE_SCP_EXTRA_ARGS=-T`` when running Ansible, or
+    * Settings ``scp_extra_args = -T`` under the ``[ssh_connection]`` config section in ``ansible.cfg``
+
+.. note:: The ``-T`` is a valid argument only for scp clients that do the filename validation. If setting that extra
```
Comment (mattclay) [inline docs/docsite/rst/reference_appendices/faq.rst]:
```suggestion
.. note:: ``-T`` is a valid argument only for scp clients that perform the filename validation. If setting that extra
```

File: docs/docsite/rst/reference_appendices/faq.rst [hunk 3]
```diff
@@ -643,6 +643,34 @@ but you can still access the original via ``hostvars``::
...
+
+How do I fix 'protocol error: filename does not match request' when fetching a file?
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+When using ``scp`` as the file transfer mechanism for SSH, it can fail with the error::
+
+    failed to transfer file to /tmp/ansible/file.txt\r\nprotocol error: filename does not match request
+
+This is due to a bug in newer releases of OpenSSH in the ``scp`` client that is used on the Ansible controller. These
+releases attempt to validate that the path of the file to fetch matches the requested requested path. This validation
```
Comment (abadger) [inline docs/docsite/rst/reference_appendices/faq.rst]:
```suggestion
releases attempt to validate that the path of the file to fetch matches the requested path. This validation
```

**Comment by abadger** [comment]
Technical portions of this look right.  Would there be a reason to list piped among the solutions as well?  (For hosts which do not enable the sftp subsystem and have the newer openssh).

Merge and backport when docs decides these are good.

File: docs/docsite/rst/reference_appendices/faq.rst [hunk 4]
```diff
@@ -643,6 +643,34 @@ but you can still access the original via ``hostvars``::
 This works for all overriden connection variables, like ``ansible_user``, ``ansible_port``, etc.
 
 
+.. _scp_protocol_error_filename:
+
+How do I fix 'protocol error: filename does not match request' when fetching a file?
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+When using ``scp`` as the file transfer mechanism for SSH, it can fail with the error::
```
Comment (acozine) [inline docs/docsite/rst/reference_appendices/faq.rst]: I'd start with the reason for the error (esp. since it's an external bug). Something like:
"Newer releases of OpenSSH have a bug in the `scp` client that can trigger this error on the Ansible controller:"

File: docs/docsite/rst/reference_appendices/faq.rst [hunk 5]
```diff
@@ -643,6 +643,34 @@ but you can still access the original via ``hostvars``::
...
+
+    failed to transfer file to /tmp/ansible/file.txt\r\nprotocol error: filename does not match request
+
+This is due to a bug in newer releases of OpenSSH in the ``scp`` client that is used on the Ansible controller. These
```
Comment (acozine) [inline docs/docsite/rst/reference_appendices/faq.rst]: Once we move the reason up, this section can be slimmed down: "The affected OpenSSH releases attempt to . . ."

File: docs/docsite/rst/reference_appendices/faq.rst [hunk 6]
```diff
@@ -643,6 +643,34 @@ but you can still access the original via ``hostvars``::
...
+releases attempt to validate that the path of the file to fetch matches the requested path. This validation
+will fail if the remote filename requires quotes to escape spaces or non-ascii chars in its path. If you come across
+this error there are a few different ways to ignore this validation, such as:
+
+* Don't use the ``scp`` transfer mechanism by
```
Comment (acozine) [inline docs/docsite/rst/reference_appendices/faq.rst]: Does this mean "avoid using `scp`, here's how" or "avoid these particular methods of using `scp`"? I'd like to rephrase this to be unambiguous and positive ("You will trigger this error if you:" or say "avoid" rather than "don't", depending on the correct meaning here).

File: docs/docsite/rst/reference_appendices/faq.rst [hunk 7]
```diff
@@ -643,6 +643,34 @@ but you can still access the original via ``hostvars``::
...
+    * Not setting ``scp_if_ssh``, by default this uses the ``smart`` mechanism which tries SFTP first,
+    * Explicitly setting the host var ``ansible_scp_if_ssh=smart`` or ``ansible_scp_if_ssh=False``,
+    * Setting the environment variable ``ANSIBLE_SCP_IF_SSH`` when running Ansible to either ``smart`` or ``False``, or
+    * Setting ``scp_if_ssh`` to either ``smart`` or ``False`` under the ``[ssh_connection]`` config section in ``ansible.cfg``
+* If SCP is needed then set the ``-T`` arg to tell the SCP client to ignore path validation by
```
Comment (acozine) [inline docs/docsite/rst/reference_appendices/faq.rst]: Active voice and address the reader here: "If you need `scp`, set the `-T` arg so the `scp` client will ignore path validation:"

File: docs/docsite/rst/reference_appendices/faq.rst [hunk 8]
```diff
@@ -643,6 +643,34 @@ but you can still access the original via ``hostvars``::
...
+    * Setting the var ``ansible_scp_extra_args=-T``,
+    * Setting the environment variable ``ANSIBLE_SCP_EXTRA_ARGS=-T`` when running Ansible, or
+    * Settings ``scp_extra_args = -T`` under the ``[ssh_connection]`` config section in ``ansible.cfg``
+
+.. note:: ``-T`` is a valid argument only for scp clients that perform the filename validation. If setting that extra
+  argument results in an invalid argument error then it shouldn't be needed at all for the Ansible controller.
```
Comment (acozine) [inline docs/docsite/rst/reference_appendices/faq.rst]: Active voice and address the reader here: "If you see an `invalid argument` error when using `-T`, then your `scp` client is not performing filename validation and will not trigger this bug."

**Review by acozine — COMMENTED** [review]
This FAQ entry has a lot of great information. I've made some suggestions to clarify and tighten up the text. I'm happy to implement them by pushing commits to this branch if you'd like - ping me when you come online.

File: docs/docsite/rst/reference_appendices/faq.rst [hunk 9]
```diff
@@ -643,6 +643,32 @@ but you can still access the original via ``hostvars``::
...
+
+    failed to transfer file to /tmp/ansible/file.txt\r\nprotocol error: filename does not match request
+
+In these releases, SCP tries to validate that the path of the file to fetch matches the requested path.
+The validation
+fails if the remote filename requires quotes to escape spaces or non-ascii characters in its path. To avoid this error:
+
+* Use SFTP instead of SCP by setting ``scp_if_ssh`` to ``smart`` (which tries SFTP first) or to ``False``. You can do this in one of four ways:
+    * Rely on the default setting, which is ``smart`` - this works if ``scp_if_ssh`` is not explicitly set anywhere
+    * Set a host variable in the ``all`` group (``ansible_scp_if_ssh=smart`` or ``ansible_scp_if_ssh=False``)
```
Comment (jborean93) [inline docs/docsite/rst/reference_appendices/faq.rst]: Doesn't need to be in the `all` group, it can even be set on the task itself.

File: docs/docsite/rst/reference_appendices/faq.rst [hunk 10]
```diff
@@ -643,6 +643,32 @@ but you can still access the original via ``hostvars``::
...
+    * Set an environment variable: ``ANSIBLE_SCP_IF_SSH=smart`` or ``ANSIBLE_SCP_IF_SSH=False``
```
Comment (jborean93) [inline docs/docsite/rst/reference_appendices/faq.rst]: This can confuse people who thing they can do

```
- module:
  environment:
    ANSIBLE_SCP_IF_SSH: smart
```

This won't work as env vars need to be set when calling Ansible itself.

**Comment by jborean93** [comment]
Backport PRs
stable-2.7 https://github.com/ansible/ansible/pull/55488
stable-2.8 https://github.com/ansible/ansible/pull/55489
